keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,981 @@
+nodes as well as the size of the pod CIDR allocated to these nodes.
+
+### User Stories
+
+#### Add more pod IPs to the cluster
+
+A user created a cluster with an initial clusterCIDR value of 10.1.0.0/20. Each
+node is assigned a /24 pod CIDR so the user could create a maximum of 16 nodes.
+However, the cluster needs to be expanded but the user does not have enough IPs
+for pods.
+
+With this enhancement, the user can now allocate an additional CIDR for pods;
+eg. 10.2.0.0/20 with the same configuration to allocate a /24 pod CIDR. This
+way, the cluster can now grow by an additional 16 nodes.
+
+#### Add nodes with higher or lower capabilities
+
+A user created a cluster with an ample sized cluster CIDR. All the initial nodes
+are of uniform capacity capable of running a maximum of 256 pods and they are
+each assigned a /24 pod CIDR. The user is planning to add more nodes to the
+system which are capable of running 500 pods. However, they cannot take
+advantage of the additional capacity because all nodes are assigned a /24 pod
+CIDR. With this enhancement the user configures a new allocation which uses the
+original cluster CIDR but allocates a /23 instead of a /24 to each node. They
+use the node selector to allocate these IPs only to the nodes with the higher
+capacity.
+
+#### Provision discontiguous ranges
+
+A user wants to create a cluster with 32 nodes each with a capacity to run 256
+pods. This means that each node needs a /24 pod CIDR range and they need a total
+range of /19. However, there aren't enough contiguous IPs in the user's network.
+They can find 4 free ranges of size /21 but no single contiguous /19 range.
+
+Using this enhancement, the user creates 4 different CIDR configurations each
+with a /21 range. The CIDR allocator allocates a /24 range from any of these /21
+ranges to the nodes and the user can now create the cluster.
+
+### Notes/Constraints/Caveats
+
+This feature does not expand the ability of the NodeIPAM controller to change
+the `Node.Spec.PodCIDRs` field. Once that field is set, either by the controller
+or a third party, it will be treated as immutable. This is particularly relevant
+in situtaitons where users start modifying or deleting the `ClusterCidrConfig`.
+Under no circumstances will the controller attempt to revoke the allocated
+CIDRs (more details on this are discussed below).
+
+### Risks and Mitigations
+
+-   Racing kube-controller-managers. If multiples of the controller are running
+    (as in a HA control plane), how do they coordinate?
+    -   The controllers will coordinate using the existing
+        kube-controller-manager leader election.
+
+## Design Details
+
+### New Resource
+
+```go
+type ClusterCIDRConfig struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec ClusterCIDRConfigSpec
+  Status ClusterCIDRConfigStatus
+}
+
+type ClusterCIDRConfigSpec {
+    // An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    // +required
+    CIDR string
```

Comment:
Any chance we could have this object provide both IP families at once? example:
```
IPv4: (optional)
  CIDR
  PerNodeMaskSize
IPv6: (optional)
  CIDR
  PerNodeMaskSize
```
Where user can define one or both families. Reasoning is:
1. Ease of configuration
2. I think with current approach we might end up with race condition for DualStack clusters i.e. the allocator will see only the first IPv4 CIDR object and assign IP from it, but there is another one with the IPv6 and it would not be used

Reply:
there is a section for dual-stack support https://github.com/kubernetes/enhancements/pull/2594/files#diff-2ca3d193cb87b069c0dbfa36429b5e380af3b02647cd4d22127cb5789b361e4dR359